pkg/gardenlet/controller/backupentry/backup_entry_migration_controller.go

```diff
@@ -0,0 +1,198 @@
+	log := r.logger.WithField("backupentry", req.String())
+
+	gardenClient, err := r.clientMap.GetClient(ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get garden client: %w", err)
+	}
+
+	backupEntry := &gardencorev1beta1.BackupEntry{}
+	if err := gardenClient.Client().Get(ctx, req.NamespacedName, backupEntry); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Infof("[BACKUPENTRY MIGRATION] Skipping because BackupEntry has been deleted")
+			return reconcile.Result{}, nil
+		}
+		log.Infof("[BACKUPENTRY MIGRATION] Unable to retrieve object from store: %+v", err)
+		return reconcile.Result{}, err
+	}
+
+	// If the backup entry is being deleted or no longer being migrated to this seed, clear migration start time and don't requeue
+	if backupEntry.DeletionTimestamp != nil || !controllerutils.BackupEntryIsBeingMigratedToThisGardenlet(backupEntry, r.config) {
+		log.Debugf("[BACKUPENTRY MIGRATION] Clearing migration start time")
+		if err := setMigrationStartTime(ctx, gardenClient.Client(), backupEntry, nil); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not clear migration start time: %w", err)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Set the migration start time if needed
+	if backupEntry.Status.MigrationStartTime == nil {
+		log.Debugf("[BACKUPENTRY MIGRATION] Setting migration start time")
+		if err := setMigrationStartTime(ctx, gardenClient.Client(), backupEntry, &metav1.Time{Time: time.Now().UTC()}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not set migration start time: %w", err)
+		}
+	}
+
+	// If the restore annotation is set or the grace period is elapsed and migration is not currently in progress,
+	// update the backup entry status to force the restoration (fallback to the "bad case" scenario)
+	log.Debugf("[BACKUPENTRY MIGRATION] Checking if the backup entry should be forcefully restored")
+	if hasForceRestoreAnnotation(backupEntry) || r.isGracePeriodElapsed(backupEntry) && !r.isMigrationInProgress(backupEntry) {
+
+		log.Infof("[BACKUPENTRY MIGRATION] Updating backup entry status to force restoration")
+		if err := updateStatusForRestore(ctx, gardenClient.Client(), backupEntry); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not update backup entry status to force restoration: %w", err)
+		}
+
+		if hasForceRestoreAnnotation(backupEntry) {
+			log.Debugf("[BACKUPENTRY MIGRATION] Removing force-restore annotation")
+			if err := removeForceRestoreAnnotation(ctx, gardenClient.Client(), backupEntry); err != nil {
+				return reconcile.Result{}, fmt.Errorf("could not remove force-restore annotation: %w", err)
+			}
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Requeue after the configured sync period
+	return reconcile.Result{RequeueAfter: r.config.Controllers.BackupEntryMigration.SyncPeriod.Duration}, nil
```

Comment:
It's not super obvious to me, why we need to requeue here. Can you explain it?

Reply:
How else we'll get the chance to actually check if the grace period has elapsed? We may not get any shoot events after the initial one for the change of `spec.seedName` for a long time (at least until the next resync or gardenlet restart, which might be much after our configured grace period). Even if we process all events, including status updates, we may still get nothing as the source seed gardenlet may be "dead".

We could requeue directly for the grace period. However, if after the grace period has elapsed the migration is still in progress, we'll still not force the restore but wait until `lastOperation.lastUpdateTime` becomes stale, so we'll still need to requeue after some arbitrarly sync period. I decided to keep it simple and always requeue for such a period.

Note that if the controller detected the shoot is no longer being migrated, or if it forced the restore (so again the shoot is no longer being migrated), we return without requeue as it's no longer needed.

Reply:
Thanks for explaining. That makes total sense.
Can you make the doc string of `SyncPeriod` more explicit and add an explanation, that it is only relevant for BackupEntries/Shoots that are currently in migration?

Reply:
Done, and also added a few more comments to the controllers themselves, hopefully this helps.